src/service/cached-xhr-impl.js

```diff
@@ -0,0 +1,124 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Cache} from '../utils/cache';
+import {Xhr} from './xhr-impl';
+import {fromClass} from '../service';
+import {removeFragment} from '../url';
+
+
+/**
+ * A wrapper around the Xhr service which caches the result of GET requests
+ *
+ * @package Visible for type.
+ * @visibleForTesting
+ */
+export class CachedXhr extends Xhr {
+
+  /**
+   * @param {!Window} win
+   * @param {number=} cacheSize
+   */
+  constructor(win, cacheSize) {
+    super(win);
+
+    /** @const {!Cache} */
+    this.cache_ = new Cache(cacheSize);
+  }
+
+  /**
+   * Attempt to cache the result of a fetch.
+   * @param {string} key The cache key
+   * @param {!./xhr-impl.FetchInitDef} init Fetch options object
+   * @param {!Promise<!JSONType>|!Promise<string>|!Promise<!Document>} fetch
+   * @return {!Promise<!JSONType>|!Promise<string>|!Promise<!Document>}
+   */
+  cacheFetch_(key, init, fetch) {
+    const isCacheable = (init.method === 'GET');
+    if (isCacheable) {
+      this.cache_.put(key, fetch);
+    }
+    return fetch;
+  }
+
+  /**
+   * Attempt to retrieve a cached fetch.
+   * @param {string} key The cache key
+   * @return {?Promise<!JSONType>|?Promise<string>|?Promise<!Document>}
+   */
+  getCachedFetch_(key) {
+    return this.cache_.get(key);
+  }
+
+  /**
+   * Fetches and caches a JSON request.
+   * @param {string} input URL
+   * @param {!./xhr-impl.FetchInitDef} init
+   * @return {!Promise<!JSONType>}
+   * @override
+   */
+  fetchJson_(input, init) {
```

Comment:
> for different Accepts header values

Don't we know that by virtue of the public method being called? Ie, we can do:

```js
fetchJSON(input, init) {
    return this.cached(input, 'json') || this.cacheFetch_(input, 'json', super.fetchJSON(input, init));
}

fetch(input, init) {
    return this.cached(input, 'generic') || this.cacheFetch_(input, 'generic', super.fetchJSON(input, init));
}

fetchText(input, init) {
    return this.cached(input, 'text') || this.cacheFetch_(input, 'text', super.fetchJSON(input, init));
}
```

`#getCacheKey` can then just concat the URL and the method to generate.